File: Python/Recursion/PermutationString.py

def permute(s,l,r):
    if(l == r):
        print(s, end = " ")
        return

    for i in range(l,r+1):
        s = swap(s,l,i)
        permute(s,l+1,r)
        s = swap(s,l,i)

def swap(s,l,i):
    s = list(s)
    temp = s[l]
    s[l] = s[i]
    s[i] = temp
    
    return "".join(s)

if __name__ == '__main__':
    permute('abc',0,2)

File: Java Programs/Recursion/PermutationString.java
// Print all permutations of the string
// abc -> {abc,acb,bac,bca,cab,cba}
// Time complexity - O(n*n!)

class PermutationString {

    static void permute(String s, int l , int r){
        if(l == r){
            System.out.print(s+ " ");
            return;
        }

        for(int i = l; i<= r; i++) {
            s = swap(s,l,i);
            permute(s,l+1,r);
            s = swap(s,l,i);
        }
    }

    static String swap(String s, int l, int i){
        char ch[] = s.toCharArray();
        char temp = ch[l];
        ch[l] = ch[i];
        ch[i] = temp;

        s = String.valueOf(ch);
        return s;
    }

    public static void main(String[] args){
        String word = "abcd";
        int l = 0;
        int r = word.length() -1;
        permute(word,l,r);
    }
}